### PR TITLE
Added a data provider for security reviews in OpenSSF

### LINKED ISSUE #524: Consider adding features for OpenSSF security reviews
See https://github.com/ossf/security-reviews

The project collects info about security reviews that have been done for open source projects. In particular, on the top of md files, there is a Yaml piece that can be easily parsed in a data provider:

https://raw.githubusercontent.com/ossf/security-reviews/main/reviews/github/madler/zlib/trail-of-bits-2016.md

Let's see if we can extract useful features for the security ratings.

### PR DESCRIPTION
**Pull request by artem-smotrakov**
Updates:

- Added a feature and a value for security reviews.
- Added an experimental data provider that fetches data about security reviews from OpenSSF project.
- Added tests.

Fixes #524

### CODE REVIEW

File: src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReview.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  /**
+   * When the review was done.
+   */
+  private final Date date;
+
+  /**
+   * Create a new review.
+   *
+   * @param subject What was reviewed.
+   * @param date Whe the review wes done.
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReview.java:28]: Typos

File: src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
+package com.sap.oss.phosphor.fosstars.model.value;
+
+import static java.util.Arrays.asList;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonGetter;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonValue;
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java:8]: Please remove the unused imports

File: src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  /**
+   * Security reviews.
+   */
+  private final Set<SecurityReview> elements = new HashSet<>();
+
+  /**
+   * Create an empty set of security reviews.
+   */
+  public SecurityReviews() {
+
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java:29]: Does not create any empty set.

File: src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  /**
+   * Returns a set of security reviews.
+   *
+   * @return A set of security reviews.
+   */
+  @JsonGetter("elements")
+  public Set<SecurityReview> elements() {
+    return new HashSet<>(elements);
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java:49]:
```suggestion
    return elements;
```

File: src/test/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviewsTest.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  @Test
+  public void testJsonSerialization() throws IOException {
+    GitHubProject project = new GitHubProject("org", "test");
+    SecurityReview firstReview = new SecurityReview(project, new Date(1));
+    SecurityReview secondReview = new SecurityReview(project, new Date(2));
+    SecurityReviews securityReviews = new SecurityReviews(firstReview, secondReview);
+
+    SecurityReviews clone = Json.read(Json.toBytes(securityReviews), SecurityReviews.class);
+    assertEquals(securityReviews, clone);
+  }
```
Comment (sourabhsparkala) [inline src/test/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviewsTest.java:40]: Seems like a duplicate of the above test.

File: src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java (hunk 4 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java:49]: This is necessary to make sure that the set can't be implicitly modified by someone.

File: src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java (hunk 3 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviews.java:29]: I am not sure that I got this statement (or question). This constructor creates an empty set. Jackson also calls it while deserialization.

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/experimental/SecurityReviewsFromOpenSSF.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    String line = reader.readLine();
+    if (!"---".equals(line)) {
+      return Optional.empty();
+    }
+
+    StringBuilder metadata = new StringBuilder();
+    do {
+      metadata.append(line).append("\n");
+      line = reader.readLine();
+    } while (line != null && !"---".equals(line));
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/data/github/experimental/SecurityReviewsFromOpenSSF.java:194]: Shouldn't this be `line != null || !"---".equals(line)`

File: src/test/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviewsTest.java (hunk 5 above)
Comment (artem-smotrakov) [reply src/test/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviewsTest.java:40]: The above test case doesn't test serialization.

File: src/main/java/com/sap/oss/phosphor/fosstars/data/github/experimental/SecurityReviewsFromOpenSSF.java (hunk 6 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/oss/phosphor/fosstars/data/github/experimental/SecurityReviewsFromOpenSSF.java:194]: Nope, it has to stop when the line is `---`.

File: src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviewsValue.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    super(feature);
+    Objects.requireNonNull(reviews, "Reviews can't be null!");
+    this.reviews = new SecurityReviews(reviews);
+  }
+
+  @Override
+  @JsonGetter("reviews")
+  public SecurityReviews get() {
+    return new SecurityReviews(reviews);
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviewsValue.java:36]: Do we need to initialize with `new` here? When it is already done in the constructor?

**Review by sourabhsparkala — APPROVED** [review]
👍

File: src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviewsValue.java (hunk 7 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/oss/phosphor/fosstars/model/value/SecurityReviewsValue.java:36]: This is necessary to make sure that the value can't be implicitly modified by the caller.

**Pull request merged**